File: plugin-rust-server/src/main/java/jetbrains/buildServer/rust/CargoRunnerRunType.java
/*
 * Copyright 2000-2016 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * See LICENSE in the project root for license information.
 */

package jetbrains.buildServer.rust;

import jetbrains.buildServer.requirements.Requirement;
import jetbrains.buildServer.serverSide.InvalidProperty;
import jetbrains.buildServer.serverSide.PropertiesProcessor;
import jetbrains.buildServer.serverSide.RunType;
import jetbrains.buildServer.serverSide.RunTypeRegistry;
import jetbrains.buildServer.web.openapi.PluginDescriptor;
import org.jetbrains.annotations.NotNull;

import java.util.*;

/**
 * Cargo runner definition.
 */
public class CargoRunnerRunType extends RunType {

    private final PluginDescriptor myPluginDescriptor;

    public CargoRunnerRunType(@NotNull final PluginDescriptor pluginDescriptor,
                              @NotNull final RunTypeRegistry runTypeRegistry) {
        myPluginDescriptor = pluginDescriptor;
        runTypeRegistry.registerRunType(this);
    }

    @NotNull
    @Override
    public String getType() {
        return CargoConstants.RUNNER_TYPE;
    }

    @NotNull
    @Override
    public String getDisplayName() {
        return CargoConstants.RUNNER_DISPLAY_NAME;
    }

    @NotNull
    @Override
    public String getDescription() {
        return CargoConstants.RUNNER_DESCRIPTION;
    }

    @Override
    public PropertiesProcessor getRunnerPropertiesProcessor() {
        return new PropertiesProcessor() {
            @Override
            public Collection<InvalidProperty> process(Map<String, String> map) {
                return Collections.emptyList();
            }
        };
    }

    @Override
    public String getEditRunnerParamsJspFilePath() {
        return myPluginDescriptor.getPluginResourcesPath("editCargoParameters.jsp");
    }

    @Override
    public String getViewRunnerParamsJspFilePath() {
        return myPluginDescriptor.getPluginResourcesPath("viewCargoParameters.jsp");
    }

    @Override
    public Map<String, String> getDefaultRunnerProperties() {
        return new HashMap<String, String>();
    }

    @NotNull
    @Override
    public String describeParameters(@NotNull Map<String, String> parameters) {
        return String.format("cargo %s", parameters.get(CargoConstants.PARAM_COMMAND));
    }

    @NotNull
    @Override
    public List<Requirement> getRunnerSpecificRequirements(@NotNull Map<String, String> runParameters) {
        return Collections.emptyList();
    }
}

File: plugin-rust-server/src/main/kotlin/jetbrains/buildServer/rust/CargoRunnerRunType.kt


package jetbrains.buildServer.rust

import jetbrains.buildServer.requirements.Requirement
import jetbrains.buildServer.requirements.RequirementType
import jetbrains.buildServer.serverSide.InvalidProperty
import jetbrains.buildServer.serverSide.PropertiesProcessor
import jetbrains.buildServer.serverSide.RunType
import jetbrains.buildServer.serverSide.RunTypeRegistry
import jetbrains.buildServer.web.openapi.PluginDescriptor

/**
 * Cargo runner definition.
 */
class CargoRunnerRunType(private val myPluginDescriptor: PluginDescriptor,
                         runTypeRegistry: RunTypeRegistry) : RunType() {

    init {
        runTypeRegistry.registerRunType(this)
    }

    override fun getType(): String {
        return CargoConstants.RUNNER_TYPE
    }

    override fun getDisplayName(): String {
        return CargoConstants.RUNNER_DISPLAY_NAME
    }

    override fun getDescription(): String {
        return CargoConstants.RUNNER_DESCRIPTION
    }

    override fun getRunnerPropertiesProcessor(): PropertiesProcessor? {
        return PropertiesProcessor {
            val errors = mutableListOf<InvalidProperty>()
            val command = it[CargoConstants.PARAM_COMMAND]
            if (command == CargoConstants.COMMAND_CUSTOM_CRATE) {
                val crate = it[CargoConstants.PARAM_CUSTOM_CRATE_COMMAND_NAME]
                if (crate.isNullOrBlank()) {
                    errors.add(InvalidProperty(CargoConstants.PARAM_CUSTOM_CRATE_COMMAND_NAME, "Crate name should not be empty"))
                }
            }
            errors
        }
    }

    override fun getEditRunnerParamsJspFilePath(): String? {
        return myPluginDescriptor.getPluginResourcesPath("editCargoParameters.jsp")
    }

    override fun getViewRunnerParamsJspFilePath(): String? {
        return myPluginDescriptor.getPluginResourcesPath("viewCargoParameters.jsp")
    }

    override fun getDefaultRunnerProperties(): Map<String, String> {
        return emptyMap()
    }

    override fun describeParameters(parameters: Map<String, String>): String {
        val command = parameters[CargoConstants.PARAM_COMMAND]
        if (command == CargoConstants.COMMAND_CUSTOM_CRATE) {
            val crate = parameters[CargoConstants.PARAM_CUSTOM_CRATE_COMMAND_NAME]
            return "cargo $crate"
        }
        return "cargo $command"
    }

    override fun getRunnerSpecificRequirements(parameters: Map<String, String>): List<Requirement> {
        val toolchainVersion = parameters[CargoConstants.PARAM_TOOLCHAIN]
        return if (toolchainVersion.isNullOrBlank()) {
            listOf(Requirement(CargoConstants.CARGO_CONFIG_PATH, null, RequirementType.EXISTS))
        } else {
            listOf(Requirement(CargoConstants.RUSTUP_CONFIG_PATH, null, RequirementType.EXISTS))
        }
    }
}